core/src/main/java/tachyon/client/tcp/TCPRemoteBlockReader.java

```diff
@@ -0,0 +1,60 @@
+package tachyon.client.tcp;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.SocketChannel;
+
+import org.apache.log4j.Logger;
+
+import tachyon.Constants;
+import tachyon.client.RemoteBlockReader;
+import tachyon.worker.DataServerMessage;
+
+public final class TCPRemoteBlockReader implements RemoteBlockReader {
+
+  private static final Logger LOG = Logger.getLogger(Constants.LOGGER_TYPE);
+
+  @Override
+  public ByteBuffer readRemoteBlock(String host, int port, long blockId, long offset, long length)
+      throws IOException {
+    InetSocketAddress address = new InetSocketAddress(host, port);
+    SocketChannel socketChannel = SocketChannel.open();
+    try {
+      socketChannel.connect(address);
+
+      LOG.info("Connected to remote machine " + address);
+      DataServerMessage sendMsg =
+          DataServerMessage.createBlockRequestMessage(blockId, offset, length);
+      while (!sendMsg.finishSending()) {
+        sendMsg.send(socketChannel);
+      }
+
+      LOG.info("Data " + blockId + " to remote machine " + address + " sent");
+
+      DataServerMessage recvMsg =
+          DataServerMessage.createBlockResponseMessage(false, blockId, null);
+      while (!recvMsg.isMessageReady()) {
+        int numRead = recvMsg.recv(socketChannel);
+        if (numRead == -1) {
+          LOG.warn("Read nothing");
+        }
+      }
+      LOG.info("Data " + blockId + " from remote machine " + address + " received");
+
+      if (!recvMsg.isMessageReady()) {
```

Comment:
I never dwelled on this code, it's the tcp code found in RemoteBlockInStream now, i just moved it to a new class.
i think that in the past the inner if in the while statment did break and thus the second if was needed, but you are right, it's not necessary anymore, i will omit it.